app/anti_virus_check.py

```diff
@@ -0,0 +1,139 @@
+                self.bound_logger.critical("OPSWAT AV does not know about this scan - the primary server may have failed")
+                raise RetryableError()
+            elif 400 <= response.status_code < 500:
+                raise BadMessageError()
+            elif 300 < response.status_code >= 500:
+                raise RetryableError()
+        else:
+            self.bound_logger.info("Response from AV OK")
+
+    def _send_for_anti_virus_check(self, filename, contents):
+        url = settings.ANTI_VIRUS_BASE_URL + "file"
+        headers = {
+            "filename": filename,
+        }
+        self._add_api_key(headers)
+
+        self.bound_logger.info("Sending for A/V scan", url=url)
+        response = requests.post(url=url, headers=headers, data=contents)
+
+        self._check_av_response(response)
+
+        self.bound_logger.info("Response received {}".format(response.text))
+        result = response.json()
+
+        if result.get("err"):
+            self.bound_logger.error("Unable to send file for anti virus scan: {}".format(result.get("err")))
+            self.bound_logger.info("Waiting before attempting again")
+            time.sleep(settings.ANTI_VIRUS_WAIT_TIME)
+            self.bound_logger.info("Return message to rabbit")
+            raise RetryableError()
+        else:
+            data_id = result.get("data_id")
+            self.bound_logger.info("File sent successfully for anti virus scan", data_id=data_id)
+            return data_id
+
+    def _get_anti_virus_result(self, data_id):
+        url = settings.ANTI_VIRUS_BASE_URL + "file/" + data_id
+        headers = {}
+        self._add_api_key(headers)
+
+        self.bound_logger.info("Getting result for A/V scan", url=url)
+
+        response = requests.get(url=url, headers=headers)
+
+        self._check_av_response(response)
+
+        self.bound_logger.info("Response from AV {}".format(response.text))
+
+        result = response.json()
+        scan_results = result.get("scan_results")
+
+        ready = False
+        safe = False
+
+        if scan_results:
+            progress_percentage = scan_results.get("progress_percentage")
+            if int(progress_percentage) == 100:
+                ready = True
+                self.bound_logger.info("Anti virus scan complete: {}".format(scan_results.get("scan_all_result_a")))
+                scan_all_result_i = scan_results.get("scan_all_result_i")
+                if int(scan_all_result_i) == 0:
+                    self.bound_logger.info("File is safe")
+                    safe = True
+                else:
+                    self.bound_logger.error("File is not safe")
+            else:
+                self.bound_logger.info("Scan not yet complete")
+        else:
+            self.bound_logger.info("Results not yet available")
+
+        return AVResult(safe=safe, ready=ready, scan_results=scan_results)
+
+    def _write_scan_report(self, av_results, filename):
```

Comment:
What is this report for?

Reply:
If the file fails the A/V this report contains all the details of which A/V products failed it and why. If we don't log it somewhere we won't have that information without going back to the OPSWAT server manually.